cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -207,8 +215,14 @@ protected void run() {
         Thread.currentThread().interrupt();
       }
     }
-    // Persist records and messageId's after all ProcessMetricsThread's complete.
-    persistRecordsMessageIds(records, topicMessageIds);
+    // Persist metricsFromAllTopics and messageId's after all ProcessMetricsThread's complete.
+    // No need to make a copy of metricsFromAllTopics and topicMessageIds because no thread is writing to them
+    if (!metricsFromAllTopics.isEmpty()) {
+      persistMetricsMessageIds(metricsFromAllTopics, topicMessageIds);
+    }
+    // Persist topicMessageIds even when metricsFromAllTopics is empty, because topicMessageIds can be updated after
+    // metricsFromAllTopics is persisted
+    persistLastMessageIds(topicMessageIds);
```

Comment:
we still need `persistLastMessageIds(topicMessageIds);` here even when the queue is empty and persisting messageId's is skipped in `persistMetricsMessageIds`